lightning/src/offers/signer.rs

```diff
@@ -0,0 +1,205 @@
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Utilities for signing offer messages and verifying metadata.
+
+use bitcoin::hashes::{Hash, HashEngine};
+use bitcoin::hashes::cmp::fixed_time_eq;
+use bitcoin::hashes::hmac::{Hmac, HmacEngine};
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::secp256k1::{KeyPair, PublicKey, Secp256k1, SecretKey, self};
+use core::convert::TryFrom;
+use core::fmt;
+use crate::ln::inbound_payment::{ExpandedKey, IV_LEN, Nonce};
+use crate::offers::merkle::TlvRecord;
+use crate::util::ser::Writeable;
+
+use crate::prelude::*;
+
+/// Message metadata which possibly is derived from [`MetadataMaterial`] such that it can be
+/// verified.
+#[derive(Clone)]
+pub(super) enum Metadata {
+	/// Metadata as parsed, supplied by the user, or derived.
+	Bytes(Vec<u8>),
+
+	/// Metadata to be derived from message contents and given material.
+	Derived(MetadataMaterial),
+
+	/// Metadata and signing pubkey to be derived from message contents and given material.
+	DerivedSigningPubkey(MetadataMaterial),
+}
+
+impl Metadata {
+	pub fn as_bytes(&self) -> Option<&Vec<u8>> {
+		match self {
+			Metadata::Bytes(bytes) => Some(bytes),
+			Metadata::Derived(_) => None,
+			Metadata::DerivedSigningPubkey(_) => None,
+		}
+	}
+
+	pub fn has_derivation_material(&self) -> bool {
+		match self {
+			Metadata::Bytes(_) => false,
+			Metadata::Derived(_) => true,
+			Metadata::DerivedSigningPubkey(_) => true,
+		}
+	}
+
+	pub fn without_keys(self) -> Self {
+		match self {
+			Metadata::Bytes(_) => self,
+			Metadata::Derived(_) => self,
+			Metadata::DerivedSigningPubkey(material) => Metadata::Derived(material),
+		}
+	}
+
+	pub fn derive_from<W: Writeable>(self, tlv_stream: W) -> (Self, Option<KeyPair>) {
+		match self {
+			Metadata::Bytes(_) => (self, None),
+			Metadata::Derived(mut metadata_material) => {
+				tlv_stream.write(&mut metadata_material.hmac).unwrap();
+				(Metadata::Bytes(metadata_material.derive_metadata()), None)
+			},
+			Metadata::DerivedSigningPubkey(mut metadata_material) => {
+				tlv_stream.write(&mut metadata_material.hmac).unwrap();
+				let (metadata, keys) = metadata_material.derive_metadata_and_keys();
+				(Metadata::Bytes(metadata), Some(keys))
+			},
+		}
+	}
+}
+
+impl Default for Metadata {
+	fn default() -> Self {
+		Metadata::Bytes(vec![])
+	}
+}
+
+impl fmt::Debug for Metadata {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			Metadata::Bytes(bytes) => bytes.fmt(f),
+			Metadata::Derived(_) => f.write_str("Derived"),
+			Metadata::DerivedSigningPubkey(_) => f.write_str("DerivedSigningPubkey"),
+		}
+	}
+}
+
+impl PartialEq for Metadata {
+	fn eq(&self, other: &Self) -> bool {
+		match self {
+			Metadata::Bytes(bytes) => if let Metadata::Bytes(other_bytes) = other {
+				bytes == other_bytes
+			} else {
+				false
+			},
+			Metadata::Derived(_) => false,
```

Comment:
Eh, this is a consequence of using `Metadata` in `Offer`, etc. It implements `PartialEq `, and so should its parts. But it never makes sense to have a derived variant once built. I think it's only needed in test assertions, so we could conditionally define it, I suppose, both here and in all the messages.

Reply:
Eh, that takes us full circle where the builder needs to hang on to the `Metadata`. Not sure if I want to re-do all that work as it would substantially affect the derivation code in the builders. Made it test-only instead.